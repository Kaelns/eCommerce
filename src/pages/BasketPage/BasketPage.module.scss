@use '@styles/abstracts/mixins';

.productsContainer {
  display: flex;
  flex-direction: column;
  gap: 2rem;
  margin-block: 2rem;

  @include mixins.media(600px) {
    align-items: center;
  }
}
