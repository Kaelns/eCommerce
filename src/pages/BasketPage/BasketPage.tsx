import { Box, Typography } from '@mui/material';
import { Title } from '@/components/typography/Title/Title';
import { useFetch } from '@/hooks/useFetch/useFetch';
import { fetchBasket } from '@/services/helpers/fetchBasket/fetchBasket';
import { LoadingFetch } from '@/components/LoadingFetch/LoadingFetch';
import { PageSkeleton } from '@/components/PageSkeleton/PageSkeleton';
import { ProductBasket } from '@/pages/BasketPage/components/ProductBasket/ProductBasket';
import { EMPTY_DATA_PRODUCTS } from '@/services/data/productsResponce/productsResponce.constants';

import styles from './BasketPage.module.scss';

export function BasketPage(): React.ReactNode {
  const { data, isLoading, error } = useFetch(fetchBasket);
  const { products, amount } = data ?? EMPTY_DATA_PRODUCTS;

  return (
    <LoadingFetch error={error} isLoading={isLoading} Skeleton={PageSkeleton} className={styles.productsContainer}>
      {products.length ? (
        <>
          <Box className={styles.productsHeader}>
            <Box>
              <Title className={styles.basket}>Product Basket</Title>
              <Typography>{amount} products</Typography>
            </Box>
          </Box>
          <Box>
            {products.map((product) => (
              <ProductBasket key={product.id} product={product} />
            ))}
          </Box>
        </>
      ) : (
        <Title>There is no products</Title>
      )}
    </LoadingFetch>
  );
}
