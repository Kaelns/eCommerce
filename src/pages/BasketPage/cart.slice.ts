import type { PayloadAction } from '@reduxjs/toolkit';
import type { Cart } from '@commercetools/platform-sdk';
import { createSlice } from '@reduxjs/toolkit';
import { MOCK_CART } from '@/services/ecommerce/constants';

const INIT_CART = {
  cart: MOCK_CART,
  discount: 0,
  isPromocode: false,
  deletionSignal: false
};

export const cartSlice = createSlice({
  name: 'cart',
  initialState: INIT_CART,
  selectors: {
    selectCart: (state) => state.cart,
    selectDiscount: (state) => state.discount,
    selectIsPromocode: (state) => state.isPromocode,
    selectDeletionSignal: (state) => state.deletionSignal
  },
  reducers: {
    setCartAction(state, action: PayloadAction<Cart>) {
      const {
        totalPrice: { centAmount },
        discountOnTotalPrice
      } = action.payload;

      const discountInMoney = discountOnTotalPrice?.discountedAmount?.centAmount ?? 0;

      state.cart = action.payload;
      state.discount = Math.round((discountInMoney / (centAmount + discountInMoney)) * 100);
    },

    resetCartAction(state) {
      state.cart = MOCK_CART;
      state.discount = 0;
      state.isPromocode = false;
      state.deletionSignal = false;
    },

    setIsPromocodeAction(state, action: PayloadAction<boolean>) {
      state.isPromocode = action.payload;
    },

    deletionSignalAction(state) {
      state.deletionSignal = !state.deletionSignal;
    }
  }
});

export const { selectCart, selectDiscount, selectIsPromocode, selectDeletionSignal } = cartSlice.selectors;
export const { setCartAction, resetCartAction, setIsPromocodeAction, deletionSignalAction } = cartSlice.actions;
