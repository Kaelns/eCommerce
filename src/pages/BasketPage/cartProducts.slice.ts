import type { ICartProducts } from '@/shared/types';
import type { PayloadAction } from '@reduxjs/toolkit';
import { cartSlice } from '@/pages/BasketPage/cart.slice';
import { createSelector, createSlice } from '@reduxjs/toolkit';
import { calculatePrice } from '@/pages/BasketPage/helpers/calculatePrice';
import { calculateQuantity } from '@/pages/BasketPage/helpers/calculateAmount';
import { convertToBasketProducts } from '@/pages/BasketPage/helpers/convertToBasketProducts';

const INIT_CART_PRODUCTS = {
  cartProducts: {} as ICartProducts,
  finalPrice: 0,
  productQuantity: 0,
  isPromocode: false
};

export const cartProductsSlice = createSlice({
  name: 'cartProducts',
  initialState: INIT_CART_PRODUCTS,
  selectors: {
    selectCartProducts: (state) => state.cartProducts,
    selectProductQuantityCart: (state) => state.productQuantity,
    selectFinalPriceCart: createSelector([(state): ICartProducts => state.cartProducts], calculatePrice),
    selectIsPromocode: (state) => state.isPromocode
  },
  reducers: {
    // setCartProductsAction(state, action: PayloadAction<ICartProducts>) {
    //   state.cartProducts = action.payload;
    //   state.productQuantity = calculateQuantity(action.payload);
    // },

    incrementQuantityAction(state, action: PayloadAction<{ id: string }>) {
      const { id } = action.payload;
      const { maxQuantity, quantity } = state.cartProducts[id];
      if (quantity < maxQuantity) {
        state.cartProducts[id].quantity += 1;
        state.productQuantity += 1;
      }
    },

    decrementQuantityAction(state, action: PayloadAction<{ id: string }>) {
      const { id } = action.payload;
      const { quantity } = state.cartProducts[id];
      if (quantity > 1) {
        state.cartProducts[id].quantity -= 1;
        state.productQuantity -= 1;
      }
    },

    setQuantityAction(state, action: PayloadAction<{ id: string; newQuantity: number }>) {
      const { id, newQuantity } = action.payload;
      const { maxQuantity, quantity } = state.cartProducts[id];
      if (newQuantity > 1 && newQuantity + quantity <= maxQuantity) {
        state.cartProducts[id].quantity = newQuantity;
        state.productQuantity = state.productQuantity - quantity + newQuantity;
      }
    },

    deleteProductAction(state, action: PayloadAction<{ id: string }>) {
      const { id } = action.payload;
      const { quantity } = state.cartProducts[id];
      state.productQuantity -= quantity;
      delete state.cartProducts[id];
    },

    setIsPromocodeAction(state, action: PayloadAction<boolean>) {
      state.isPromocode = action.payload;
    }
  },
  extraReducers: (builder) => {
    builder.addCase(cartSlice.actions.setCartAction, (state, action) => {
      state.cartProducts = convertToBasketProducts(action.payload.lineItems);
      state.productQuantity = calculateQuantity(action.payload.lineItems);
    });
  }
});

export const { selectCartProducts, selectFinalPriceCart, selectProductQuantityCart } = cartProductsSlice.selectors;
