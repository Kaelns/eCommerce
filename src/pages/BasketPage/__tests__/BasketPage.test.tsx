import { BasketPage } from '@/pages/BasketPage/BasketPage';
import { render } from '@testing-library/react';

describe('Given BasketPage component: ', () => {
  it('when rendered, should match snapshot', () => {
    const { asFragment } = render(<BasketPage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
