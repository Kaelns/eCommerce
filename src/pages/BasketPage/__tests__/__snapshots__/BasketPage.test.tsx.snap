// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given BasketPage component:  > when rendered, should match snapshot 1`] = `
<DocumentFragment>
  <h5
    class="MuiTypography-root MuiTypography-h5 _basket_01fa68 css-bezmlo-MuiTypography-root"
  >
    Basket Page
  </h5>
</DocumentFragment>
`;
